### PR TITLE
start integration testing and some fixes

### PR DESCRIPTION
**Pull request by bennycio**
Changes:

- **Added** integration testing for indexer covering a lot of cases, more cases to come
- **Changed** some code causing refresh issues

### CODE REVIEW

**Review by jarrel-b — COMMENTED** [review]
Awesome job!! Excited to have this stuff in place.

My ideas for tests in the future
* testing invariants: tokens already indexed before should be untouched, tokenID + contract + tokenType are never null
* what happens when running the same block range a second time
* confirming what happens for tokens that already existed (upserts)
* checking the other fields (token type, token id, name, etc)
* testing different types of transfers
* testing contracts

Would we want to add this to our CICD later on?

File: service/multichain/multichain.go [hunk 1]
```diff
@@ -426,11 +426,11 @@ func (d *Provider) RefreshToken(ctx context.Context, ti persist.TokenIdentifiers
 			}
 		}
 
-		tokens, contracts, err := provider.GetTokensByTokenIdentifiers(ctx, id)
-		if err != nil {
-			return err
-		}
 		if i == 0 {
+			tokens, contracts, err := provider.GetTokensByTokenIdentifiers(ctx, id)
```
[the remark this answers is not in the document]
Comment (bennycio) [reply service/multichain/multichain.go:430]: Probably not, but I do think we want to have a second PR that will actually use the results from the other providers and go through the fallback system. We can leave this here until then.

File: indexer/t__utils_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	return assert.New(t), db
+}
+
+func newMockIndexer(db *sql.DB) *indexer {
+	start := uint64(0)
+	end := uint64(100)
+	i := newIndexer(nil, nil, nil, nil, postgres.NewTokenRepository(db), postgres.NewContractRepository(db), persist.ChainETH, []eventHash{transferBatchEventHash, transferEventHash, transferSingleEventHash}, func(ctx context.Context, curBlock, nextBlock *big.Int, topics [][]common.Hash) ([]types.Log, error) {
+		if curBlock.Uint64() == 0 {
+			return []types.Log{
+				{
```
[the remark this answers is not in the document]
Comment (bennycio) [reply indexer/t__utils_test.go:51]: for sure!

File: indexer/t__integration_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	ethClient := rpc.NewEthClient()
+	ipfsShell := rpc.NewIPFSShell()
+	arweaveClient := rpc.NewArweaveClient()
+	stg := newStorageClient(ctx)
+
+	uri := testBase64JSONURI(ctx, ethClient, a)
```
[the remark this answers is not in the document]
Comment (bennycio) [reply indexer/t__integration_test.go:38]: If I understand this question correctly, the results of the ran indexer are what are being used in these test cases. The indexer indexes the logs for the various types of tokens at the start of the test, then we run metadata handling for those just indexed tokens.

**Pull request merged**